vowpalwabbit/cbify.cc

```diff
@@ -386,35 +686,101 @@ void finish_multiline_example(vw& all, cbify&, multi_ex& ec_seq)
 base_learner* cbify_setup(options_i& options, vw& all)
 {
   uint32_t num_actions = 0;
+  uint32_t cb_continuous_num_actions = 0;
   auto data = scoped_calloc_or_throw<cbify>();
   bool use_cs;
+  bool use_reg;  // todo: check
+  bool use_discrete;
 
   option_group_definition new_options("Make Multiclass into Contextual Bandit");
   new_options
       .add(make_option("cbify", num_actions)
                .keep()
                .help("Convert multiclass on <k> classes into a contextual bandit problem"))
-      .add(make_option("cbify_cs", use_cs).help("consume cost-sensitive classification examples instead of multiclass"))
+      .add(make_option("cbify_cs", use_cs).help("Consume cost-sensitive classification examples instead of multiclass"))
+      .add(make_option("cbify_reg", use_reg)
+               .help("Consume regression examples instead of multiclass and cost sensitive"))
+      .add(make_option("cats", cb_continuous_num_actions)
+               .default_value(0)
+               .keep()
+               .help("Continuous action tree with smoothing"))
+      .add(make_option("cb_discrete", use_discrete)
+               .keep()
+               .help("Discretizes continuous space and adds cb_explore as option"))
+      .add(make_option("min_value", data->regression_data.min_value).keep().help("Minimum continuous value"))
+      .add(make_option("max_value", data->regression_data.max_value).keep().help("Maximum continuous value"))
+      .add(make_option("loss_option", data->regression_data.loss_option)
+               .default_value(0)
+               .help("loss options for regression - 0:squared, 1:absolute, 2:0/1"))
+      .add(make_option("loss_report", data->regression_data.loss_report)
+               .default_value(0)
+               .help("loss report option - 0:normalized, 1:denormalized"))
+      .add(make_option("loss_01_ratio", data->regression_data.loss_01_ratio)
+               .default_value(0.1f)
+               .help("ratio of zero loss for 0/1 loss"))
       .add(make_option("loss0", data->loss0).default_value(0.f).help("loss for correct label"))
       .add(make_option("loss1", data->loss1).default_value(1.f).help("loss for incorrect label"));
+
   options.add_and_parse(new_options);
 
   if (!options.was_supplied("cbify"))
     return nullptr;
 
+  data->regression_data.num_actions = num_actions;
   data->use_adf = options.was_supplied("cb_explore_adf");
   data->app_seed = uniform_hash("vw", 2, 0);
   data->a_s = v_init<action_score>();
   data->all = &all;
 
   if (data->use_adf)
-    init_adf_data(*data, num_actions);
+  { init_adf_data(*data.get(), num_actions); }
 
-  if (!options.was_supplied("cb_explore") && !data->use_adf)
+  if (use_reg)  // todo: check: we need more options passed to pmf_to_pdf
```

Comment:
Can you address this TODO before merging?

Reply:
Done.